servers/physics_3d/step_3d_sw.h

```diff
@@ -33,13 +33,18 @@
 
 #include "space_3d_sw.h"
 
+#include "core/templates/local_vector.h"
+
 class Step3DSW {
 	uint64_t _step;
 
-	void _populate_island(Body3DSW *p_body, Body3DSW **p_island, Constraint3DSW **p_constraint_island);
-	void _setup_island(Constraint3DSW *p_island, real_t p_delta);
-	void _solve_island(Constraint3DSW *p_island, int p_iterations, real_t p_delta);
-	void _check_suspend(Body3DSW *p_island, real_t p_delta);
+	LocalVector<LocalVector<Body3DSW *>> body_islands;
+	LocalVector<LocalVector<Constraint3DSW *>> constraint_islands;
```

Comment:
The local vector doesn't deallocate the memory when you remove an object, so it works really well when you use it a lot.

When you copy a vector, you lost any pre-allocated memory: https://github.com/godotengine/godot/blob/master/core/templates/local_vector.h#L238-L251
By calling `constraint_islands.resize(island_count);`, you may lose the pre-allocated memory for the sub vectors. Each time you add a new object to the isle, since the vector is ""new"", it may need to re-allocate the memory. If you can avoid that, it would be better.

Have you considered to create a `LocalVector` pool, where you reuse the vectors instead? A naive solution may be use `new`:
```diff
- LocalVector<LocalVector<>>
+ LocalVector<LocalVector<> *>
```

Reply:
Using a `LocalVector` pool is a great idea, I'll see if I can update to add that.

Reply:
After testing this again, it doesn't look like any copy happens in the situation where `constraint_islands.resize(island_count);` has to re-allocate. It just does a `memrealloc` and the sub vectors are still keeping their capacity intact:
https://github.com/godotengine/godot/blob/b4b7c97d3839170180e4f3f7c2bd3179b374057a/core/templates/local_vector.h#L141-L150

Reply:
That's correct for the vectors that are not removed, but in that function we have:

https://github.com/godotengine/godot/blob/master/core/templates/local_vector.h#L136
```cpp
for (U i = p_size; i < count; i++) {
	data[i].~T();
}
```
and
https://github.com/godotengine/godot/blob/master/core/templates/local_vector.h#L153
```
for (U i = count; i < p_size; i++) {
	memnew_placement(&data[i], T);
}
```
That take care to remove and allocate the new sub vectors, in this case.

This mean that if in frame 1 you have 10 islands, the following frame it drops to 5, then it reach 10 again basically you have to reallocate 5 vectors again. Which is a waste.
If, it doesn't take too much time, you could research a way to avoid that: maybe by never resizing the main vector, but using an external variable to track the "active" isles.

Reply:
Ah yeah, it shouldn't be too hard to only grow the vector's size and keep track of the active islands as you suggested. I'll test in a scenario with many islands to see if it helps when I have a chance.

Reply:
@AndreaCatania Just as a follow-up, I've checked this scenario, and I didn't remember but the logic is already that it only grows. Basically the main vector can only resize following this logic:
```
if (constraint_islands.size() < island_count) {
	constraint_islands.resize(island_count);
}
```
Which will prevent any previous sub-vectors from being deallocated.